### PR TITLE
Add LinkGenerator extensions for MVC

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        /// <param name="httpContext">The <see cref="HttpContext"/> associated with the current request.</param>
+        /// <param name="action">
+        /// The action name. Used to resolve endpoints. Optional. If <c>null</c> is provided, the current action route value
+        /// will be used.
+        /// </param>
+        /// <param name="controller">
+        /// The controller name. Used to resolve endpoints. Optional. If <c>null</c> is provided, the current controller route value
+        /// will be used.
+        /// </param>
+        /// <param name="values">The route values. Optional. Used to resolve endpoints and expand parameters in the route template. Optional.</param>
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: Optional: removing one or more optional͓̦s. Op͚tio̫͓̬nal.

**O̺͖̭͕p͚t̳̘ͅi̲̲̻̤͍̯̳͡o̫͓̬ṉ̠a͚l͓̦!**

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        /// <param name="pathBase">
+        /// An optional URI path base. Prepended to the path in the resulting URI. If not provided, the value of <see cref="HttpRequest.PathBase"/> will be used.
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: nit: Some params start with "an optional XXX" and others are all like "The XXX. Optional."

Pick one for consistency?

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        /// </param>
+        /// <param name="fragment">An optional URI fragment. Appended to the resulting URI.</param>
+        /// <param name="options">
+        /// An optional <see cref="LinkOptions"/>. Settings on provided object override the settings with matching
+        /// names from <c>RouteOptions</c>.
+        /// </param>
+        /// <returns>A URI with an absolute path, or <c>null</c>.</returns>
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: Say when null will be returned here, in the summary, or both

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                pathBase,
+                fragment,
+                options);
+        }
+
+        /// <summary>
+        /// Generates an absolute URI based on the provided values.
+        /// </summary>
+        /// <param name="generator">The <see cref="LinkGenerator"/>.</param>
+        /// <param name="action">The action name. Used to resolve endpoints. </param>
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: nit: trailing space

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            var address = CreateAddress(httpContext: null, action, controller, values);
+            return generator.GetUriByAddress<RouteValuesAddress>(address, address.ExplicitValues, scheme, host, pathBase, fragment, options);
+        }
+
+        /// <summary>
+        /// Gets a <see cref="LinkGenerationTemplate"/> based on the provided <paramref name="action"/>, <paramref name="controller"/>, and <paramref name="values"/>.
+        /// </summary>
+        /// <param name="generator">The <see cref="LinkGenerator"/>.</param>
+        /// <param name="action">The action name. Used to resolve endpoints.</param>
+        /// <param name="controller"> The controller name. Used to resolve endpoints.</param>
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: nit: prefix space

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Routing/ControllerLinkGeneratorExtensions.cs]: Sure, I'll do the latter.

**Comment by rynowak** [comment]
> What about area?

Adding area to these directly is something we should consider. I'm going to merge this in and then spend a little time trying to figure out what that would look like.

**Pull request merged**